types/lodash/ts3.1/common/array.d.ts

```diff
@@ -0,0 +1,2090 @@
+        /**
+         * @see _.fill
+         */
+        fill<T, U>(array: U[] | null | undefined, value: T, start?: number, end?: number): Array<T | U>;
+        /**
+         * @see _.fill
+         */
+        fill<T, U>(array: List<U> | null | undefined, value: T, start?: number, end?: number): List<T | U>;
+    }
+    interface Collection<T> {
+        /**
+         * @see _.fill
+         */
+        fill<U>(value: U, start?: number, end?: number): Collection<T | U>;
+    }
+    interface CollectionChain<T> {
+        /**
+         * @see _.fill
+         */
+        fill<U>(value: U, start?: number, end?: number): CollectionChain<T | U>;
+    }
+    interface LoDashStatic {
+        /**
+         * This method is like _.find except that it returns the index of the first element predicate returns truthy
+         * for instead of the element itself.
+         *
+         * @param array The array to search.
+         * @param predicate The function invoked per iteration.
+         * @param fromIndex The index to search from.
+         * @return Returns the index of the found element, else -1.
+         */
+        findIndex<T>(array: List<T> | null | undefined, predicate?: ListIterateeCustom<T, boolean>, fromIndex?: number): number;
+    }
+    interface Collection<T> {
+        /**
+         * @see _.findIndex
+         */
+        findIndex(predicate?: ListIterateeCustom<T, boolean>, fromIndex?: number): number;
+    }
+    interface CollectionChain<T> {
+        /**
+         * @see _.findIndex
+         */
+        findIndex(predicate?: ListIterateeCustom<T, boolean>, fromIndex?: number): PrimitiveChain<number>;
+    }
+    interface LoDashStatic {
+        /**
+         * This method is like _.findIndex except that it iterates over elements of collection from right to left.
+         *
+         * @param array The array to search.
+         * @param predicate The function invoked per iteration.
+         * @param fromIndex The index to search from.
+         * @return Returns the index of the found element, else -1.
+         */
+        findLastIndex<T>(array: List<T> | null | undefined, predicate?: ListIterateeCustom<T, boolean>, fromIndex?: number): number;
+    }
+    interface Collection<T> {
+        /**
+         * @see _.findLastIndex
+         */
+        findLastIndex(predicate?: ListIterateeCustom<T, boolean>, fromIndex?: number): number;
+    }
+    interface CollectionChain<T> {
+        /**
+         * @see _.findLastIndex
+         */
+        findLastIndex(predicate?: ListIterateeCustom<T, boolean>, fromIndex?: number): PrimitiveChain<number>;
+    }
+    interface LoDashStatic {
+        /**
+         * @see _.head
+         */
+        first: LoDashStatic["head"];
+    }
+    interface String {
+        /**
+         * @see _.first
+         */
+        first(): string | undefined;
+    }
+    interface StringChain {
+        /**
+         * @see _.first
+         */
+        first(): StringChain;
```

Comment:
Here are the methods that might return undefined (as found by a simple grep, not probably missing a few):

* first/last
* sample
* find[Last]Key
* min/max[By]
* pop/shift
* reduce[Right]
* get
* nth

So, 15 in total.
I'll introduce `StringNullableChain` and see how it affects performance. If the hit of introducing more overloads to `chain/ExpChain` and another set of methods is too great then I'll hold off on this change. I don't think losing nullability in explicit chains is *that* bad because `_.chain` is relatively rare for lodash.

Reply:
Actually push, pop, min, max, sample, findKey, nth and reduce already notate nullability on the collection and object methods, they just need to use ExpChain to get the correct. get does not, but I don't think it's worthwhile to introduce an CollectionNullableChain for one method.

I'll push a commit with StringNullableChain since improves nullability tracking for strings a decent amount.